### PR TITLE
Move Sega PCM Bank scheme to device address map

### PR DESCRIPTION
**Pull request by cam900**
Implement segapcm interface to vgmplay

### CODE REVIEW

File: src/mame/drivers/segaxbd.cpp [hunk 1]
```diff
@@ -316,9 +316,25 @@ segaxbd_state::segaxbd_state(const machine_config &mconfig, device_type type, co
...
+	uint8_t *src    =   memregion(region)->base();
+	std::vector<uint8_t> buf(len);
+
+	for (int i = 0; i < len / 2; i += 0x10000)
+	{
+		memcpy(&buf[(i * 0x20000)],           src[i * 0x10000], 0x10000);
+		memcpy(&buf[(i * 0x20000) + 0x10000], src[i * 0x10000], 0x10000);
+	}
+	memcpy(&src[0], &buf[0], len);
+}
```
Comment (cuavas) [inline src/mame/drivers/segaxbd.cpp]: This screams that it's a hack.  Are you sure you can't do this properly with address maps?

File: src/mame/drivers/vgmplay.cpp [hunk 2]
```diff
@@ -1159,10 +1172,15 @@ void vgmplay_state::machine_start()
 			m_ym2612->set_unscaled_clock(r32(0x2c));
 		if(version >= 0x110 && r32(0x30))
 			m_ym2151->set_unscaled_clock(r32(0x30));
-
+			
 		if(version >= 0x151 && r32(0x38))
 			m_segapcm->set_unscaled_clock(r32(0x38));
 
+		if(version >= 0x151 && r32(0x3c)) {
```
Comment (cuavas) [inline src/mame/drivers/vgmplay.cpp]: This doesn't match the style of the surrounding code.  The statement immediately below is Allman style.

File: src/devices/sound/segapcm.cpp [hunk 3]
```diff
@@ -20,12 +21,9 @@ segapcm_device::segapcm_device(const machine_config &mconfig, const char *tag, d
 	: device_t(mconfig, SEGAPCM, tag, owner, clock),
 		device_sound_interface(mconfig, *this),
 		device_rom_interface(mconfig, *this, 21),
-		m_ram(nullptr),
-		m_bank(0),
-		m_bankshift(0),
-		m_bankmask(0),
 		m_stream(nullptr)
 {
+	std::fill(std::begin(m_ram), std::end(m_ram), 0);
```
Comment (cuavas) [inline src/devices/sound/segapcm.cpp]: Zero-filling `m_ram` on construction is wasted work when it's going to be filled with 0xff at start time.  But that aside, why did you change it from being dynamically allocated to being a member array?  Making it dynamically allocated at start time can improve the performance of `-validate`, `-listxml`, `-romident` and other verbs that need to instantiate devices but not start them.  Making it a member and filling it at construction increases the work required to construct an instance of the device.

File: src/devices/sound/segapcm.cpp [hunk 4]
```diff
@@ -35,10 +35,20 @@ void segapcm_device::device_start()
 {
 	m_stream = stream_alloc(0, 2, clock() / 128);
 	
-	std::fill(std::begin(m_ram), std::end(m_ram), 0xff);
+	m_ram = make_unique_clear<uint8_t[]>(0x800);
```
Comment (cuavas) [inline src/devices/sound/segapcm.cpp]: Why are you using `make_unique_clear` here when you plan to fill it with 0xff at reset time anyway?  It's wasted work to zero-fill something when you plan to completely overwrite it anyway.

**Comment by cam900** [comment]
I close this Because I uploaded similar PR.

**Pull request closed without merging**